Implements fandom tag pages and fixes Dashboard issues

SigmasonicX opened the pull request:
Addresses ticket #636 and #649

## Description
When you click fandom tag, you go to page that describes the tag and lists stories with that tag.

## Changes
- Creates page for each fandom tag. This includes a link to its parent if available, the description, links to its children if available, and all works with this fandom tag. This does not include all works with its children as fandom tags.
- Changes the tag listings in Dashboard, My Stuff, work cards, and content boxes to have links.
- All changes from #650

## Areas Affected
- [ ] Site Navigation
- [ ] Home
- [ ] Sign In/Out
- [ ] Registration
- [ ] Settings

.
- [ ] Profile
- [x] My Stuff (main page)
- [x] Editor
- [x] Browse
- [x] Work Card

.
- [x] Work Page
- [ ] Blog Page
- [ ] Collections
- [ ] Comments
- [ ] Other Pages

.
- [ ] Account Authentication
- [x] Dashboard
- [ ] Mobile

SigmasonicX committed:
Create blank tag page and link to it in work card tags

SigmasonicX committed:
Adds links to tags in content-box

SigmasonicX committed:
Adds tag links to content preview

SigmasonicX committed:
Fixes deprecated uses of combineLatest()

SigmasonicX committed:
Merge branch 'main' into sigmasonicx/tag_pages

SigmasonicX committed:
Adds parameters to tag page URL

SigmasonicX committed:
Fetches tag information, including children

SigmasonicX committed:
Sets tag page title, displays tag parent and links it
- Removes extra spaces after ng-container tag

SigmasonicX committed:
Merge branch 'main' into sigmasonicx/tag_pages

SigmasonicX committed:
Addresses two tags problems
- Handles tag names and descriptions as HTML when displaying
- In tag management page, moves description to expanded tag box
- Removes redundant mat-expansion scss code

SigmasonicX committed:
Handles tag HTML in child tag display and tag form

SigmasonicX committed:
Keeps current tag box open after making tag changes

SigmasonicX committed:
Checks for tag name duplication only between parent tags, and between child tags of the same parent

SigmasonicX committed:
Merge branch 'sigmasonicx/tag_pages' into sigmasonicx/fix_tag_problems

SigmasonicX committed:
Merge pull request #650 from OffprintStudios/sigmasonicx/fix_tag_problems

Fix tag problems (PR for tag_pages branch)

SigmasonicX committed:
Adds tag page links to Dashboard tag management page

SigmasonicX committed:
Creates shared htmlDecode() method and decodes HTML in tag pages
-

SigmasonicX committed:
On tag page, fetches all works with that tag, and lists children
- Makes page scrollable and adjusts display
- Attempts to implement multiple pages, but page selection doesn't work
- Very bottom of page is cut off

SigmasonicX committed:
Removes redundant code from work-card.component related to tags
- Also removes extra spaces

SigmasonicX committed:
Implements page param checking, and removes no results text

SigmasonicX commented on apps/api/src/app/controllers/search/search.controller.ts:
```diff
@@ -51,4 +51,14 @@ export class SearchController {
+        @Cookies('contentFilter') contentFilter: ContentFilter
+    ): Promise<PaginateResult<ContentModel>> {
+        return await this.searchService.searchFandomTagContent(tagId, pageNum, contentFilter);
+    }
```
Added method to get fandom tag content as part of search; in the future, this will be consolidated into a more advanced general search method, since #637 will require being able to search by tag anyway.

pingzing replied:
nitpick: The name of this method confuses me a little--does it get the content of a fandom tag, or the set of all content _with_ a particular fandom tag? If the latter, maybe `getContentByFandomTag`?

SigmasonicX commented on apps/api/src/app/services/search/search.service.ts:
```diff
@@ -12,19 +12,28 @@ import { ContentKind, ContentModel } from '@dragonfish/shared/models/content';
+    INITIAL_PAGE = 1;
+    INITIAL_MAX_PER_PAGE = 6;
+    MAX_PER_PAGE = 15;
+
```
As these numbers are repeated often throughout this class, I made them fields.

pingzing replied:
always a fan. 👍 
nitpick: make 'em const too, to clarify intent

pingzing replied:
oh right, class variables can't be const. I keep forgetting that...

SigmasonicX commented on apps/bettafish/src/app/app-routing.module.ts:
```diff
@@ -33,6 +34,7 @@ const routes: Routes = [
     ...ContentViewRoutes,
     ...SettingsRoutes,
     ...RegistrationRoutes,
+    ...TagRoutes,
```
To support going to the tag pages.

SigmasonicX commented on apps/bettafish/src/app/components/content/content-box/content-box.component.html:
```diff
@@ -60,8 +60,18 @@ <h2 class='text-white text-4xl font-medium hover:cursor-pointer'>
+                                            , 
+                                        </ng-container>
+                                    </ng-container>
+                                </span>
```
Adds links to tags, and handles HTML conversion like \&amp; -> &

SigmasonicX commented on apps/bettafish/src/app/components/content/content-box/content-box.component.scss:
```diff
@@ -37,4 +37,13 @@ div.content-box {
+                background: var(--accent-dark);
+            }
+        }
+    }
```
Makes tags visible as links

SigmasonicX commented on apps/bettafish/src/app/pages/content-views/news-page/news-page.component.ts:
```diff
@@ -26,7 +26,7 @@ export class NewsPageComponent implements OnInit {
 
     ngOnInit(): void {
-        combineLatest(this.viewState.currContent$, this.route.queryParamMap)
+        combineLatest([this.viewState.currContent$, this.route.queryParamMap])
```
Handles warning by converting parameters to array; separate parameters in this method is deprecated

SigmasonicX commented on apps/bettafish/src/app/pages/tags/tag-page.component.html:
```diff
@@ -0,0 +1,56 @@
+<dragonfish-topbar></dragonfish-topbar>
```
This and the TS file below are the main additions

SigmasonicX commented on apps/bettafish/src/material.scss:
```diff
@@ -86,35 +86,6 @@ button.mat-menu-item {
-    color: var(--text-color) !important;
-    border: 1px solid var(--borders) !important;
-}
-
```
Removes code that is overwritten right below it.

SigmasonicX commented on libs/api/database/src/lib/content/schemas/content.schema.ts:
```diff
@@ -79,7 +79,7 @@ export class ContentDocument extends Document implements ContentModel {
         },
     }] })
-    tags?: TagsModel[];
+    tags?: string[] | TagsModel[];
```
Needed to check tags with string queries

SigmasonicX commented on libs/api/database/src/lib/content/stores/tags.store.ts:
```diff
@@ -153,10 +172,20 @@ export class TagsStore {
+            const tagWithSameName = await this.findParentTagByName(newTag.name)
+            if (tagWithSameName) {
+                throw new BadRequestException("There is already a parent tag with this name.");
+            }
```
Improves name duplication check when creating/editing tags by just checking on the same level

SigmasonicX commented on libs/client/dashboard/src/lib/pages/tags-management/tags-management.component.html:
```diff
@@ -25,15 +25,17 @@ <h3>Looks like there something missing here...</h3>
                 <mat-accordion>
                     <ng-container *ngFor="let tag of tags">
-                        <mat-expansion-panel>
+                        <mat-expansion-panel [expanded]="(tag._id === openPanelId)? true : false" (opened)="onOpen(tag._id)">
```
Keeps the tag panel open after making changes, as opposed to closing it

SigmasonicX commented on libs/client/repository/src/lib/tags/tags.service.ts:
```diff
@@ -36,7 +36,9 @@ export class TagsService {
-            tap(() => {}),
+            tap((tagTree) => {
+                this.tagsStore.set([tagTree]);
+            }),
```
Used in tag pages to get children

SigmasonicX commented on libs/shared/src/lib/functions/html-decode.ts:
```diff
@@ -0,0 +1,4 @@
+export function htmlDecode(input: string): string {
+    var doc = new DOMParser().parseFromString(input, "text/html");
+    return doc.documentElement.textContent;
+}
```
Method used to handle \&amp; -> & and similar HTML processing outside of HTML files

pingzing replied:
oh good, hopefully we can stop having amps in places where people don't want them. 👍

SigmasonicX committed:
Fixes how tags are displayed in My Stuff's prose form
- Also replaces div use with span in Dashboard

pingzing commented on apps/bettafish/src/app/pages/tags/tag-page.component.html:
```diff
@@ -1,49 +1,56 @@
+            <ng-container *ngIf="tags.length === 0; else notEmpty">
+                <div class="empty">
+                    <h3>Looks like there something missing here...</h3>
+                    <p>If you think this is an error, contact your system administrator.</p>
```
nitpick: I know this is a standard error (and I think we use it everywhere) but we may want to consider changing this to something more appropriate (i.e. "please let us know in Discord" or something)

SigmasonicX replied:
I think updating that is for another PR

pingzing commented on apps/bettafish/src/app/pages/tags/tag-page.component.html:
```diff
@@ -1,49 +1,56 @@
+                        <ng-container *ngIf="tags[0].parent">
+                            <a
+                                [routerLink]="['/tag', tags[0].parent._id, tags[0].parent.name | slugify]"
+                                [innerHtml]="tags[0].parent.name | safeHtml"
```
nitpick: Angular's recommend styles tend to avoid `innerHtml` in favor of `{{ interpolation }}`. I don't think we have a precedent in the codebase, but as a _personal_ preference, I find interpolation a bit more readable, i.e.
```html
<a [routerLink]="['/tag', tags[0].parent._id, tags[0].parent.name | slugify]">
  {{ tags[0].parent.name | safeHtml }}
</a>
```

As an aside, you shouldn't need to run a tag name through `safeHtml`, as tag names shouldn't contain HTML.

...that said, do we have anything preventing a malicious moderator from creating a tag name with HTML tags in it?

SigmasonicX replied:
That's how it was before, and it didn't process the HTML correctly

SigmasonicX replied:
Regarding creating tags with HTML tags in the name, we run sanitize() on the names and descriptions; this is just being doubly safe.

pingzing commented on apps/bettafish/src/app/pages/tags/tag-page.component.html:
```diff
@@ -1,49 +1,56 @@
+                            ></a>
+                            &nbsp;— 
+                        </ng-container>
+                        <span [innerHtml]="tags[0].name | safeHtml"></span>
```
nitpick: ditto the `innerHtml` versus interpolation here. etc etc for the rest of the file

pingzing commented on apps/bettafish/src/app/pages/tags/tag-page.component.html:
```diff
@@ -1,49 +1,56 @@
-    </ng-container>
-</ng-template>
+                    <p>
+                        <span [innerHtml]="tags[0].desc | safeHtml"></span>
```
ditto desc-as-html but oh no do we sanitize that in the backend either

pingzing commented on apps/bettafish/src/app/pages/tags/tag-page.component.ts:
```diff
@@ -1,25 +1,73 @@
+            else {
+                setTwoPartTitle(htmlDecode(tagsTree.name));
+            }
+        });
```
We can save some network traffic here. As currently written, we're re-fetching tag data on every page change, but we can cache it, because it's unlikely to change. You'll want to call it once, then just use the cached value, i.e. something like this:

```TypeScript
import { lastValueFrom } from 'rxjs';

// As a class-level variable:
private cachedTags: TagsTree = null;

private async fetchData(tagId: string, pageNum: number): Promise<void> { // now that this is async, remember to make the callback passed to .subscribe() up above async as well)
  if (!this.cachedTags) {
    this.cachedTags = await lastValueFrom(this.tagsService.fetchDescendants(tagId));
    if (this.cachedTags.parent) {
      // etc
    }
    else {
     // etc
    } 
  }
// Though if you wanted, you could turn this into an awaitable promise with lastValueFrom() too!
this.network.searchFandomTagContent(tagId, pageNum).subscribe((results) => {
  this.searchResults = results;
});
}

```

SigmasonicX replied:
Since the tag tree is already stored in the LocalTagsStore and accessed with TagsQuery, there isn't a need to save it as a field. When clicking around, I noticed some odd behavior that I accounted for in the new code.

pingzing replied:
oh right, I keep forgetting we have a Store layer that already caches for us between pages and the network service...

pingzing reviewed (changes requested):
The only change I actually care about is the caching of tags on the `tag-page-component.ts`. Everything else is minor stuff you can choose to ignore if you'd lile.

SigmasonicX committed:
Renames all content-by-tag fetching methods to getContentByFandomTag() and
- Makes SearchService fields readonly
- Sorts getContentByFandomTag() results by published on

SigmasonicX committed:
Fixes page controls on tag screen and
- Properly fetches page query parameter from URL
- Only fetches tag information when URL has a new tag ID
- To address unusual behavior when clicking between child and parent tags, where it stays on ex: page number 3 if there is no page parameter, sets page number to 1 if there isn't a query parameter for it
- Prevents topbar from pushing contents past scrollable area

pingzing commented on apps/bettafish/src/app/pages/tags/tag-page.component.ts:
```diff
@@ -33,24 +33,30 @@ export class TagPageComponent implements OnInit {
-        this.network.searchFandomTagContent(tagId, pageNum).subscribe((results) => {
+        if (this.tagId != tagId) {
+            this.tagId = tagId;
+            this.tagsService.fetchDescendants(tagId).subscribe((tagsTree) => {
+                if (tagsTree.parent) {
+                    setTwoPartTitle(htmlDecode((tagsTree.parent as TagsModel).name + " — " + tagsTree.name));
+                }
+                else {
+                    setTwoPartTitle(htmlDecode(tagsTree.name));
+                }
+            });
```
If I wanted to _really_ pick some nits, we're setting the title here on every page change when we don't need to. But if someone pointed that out to me in my PR, I'd ignore it =P

SigmasonicX replied:
Actually no, that only happens if the tag ID in the URL doesn't match what it was before, that is to say if the user moves to a different tag.

pingzing replied:
oh, duh, I even read that part too and went "ah yes, smart". ...then immediately forgot about it
